text/0695-default-args.md

```diff
@@ -0,0 +1,556 @@
+Ember ships with.
+
+#### Extending `templateOnly`
+
+Today, users can define a template only component with the
+[`templateOnly`](https://github.com/emberjs/ember.js/blob/v3.23.1/packages/%40ember/component/template-only.ts#L35)
+API. This API is generally meant to be a compile target which users do not
+actually write, and as such it is the perfect place for us to add our
+`defaultArgs` integration. Since users are not meant to write this directly, it
+does not establish an opinionated high-level API.
+
+Currently, this function optionally receives the module name of the associated
+component as its first parameter. We propose extending this API so that it can
+receive an options object as the first parameter with the following interface:
+
+```ts
+function templateOnly(options: {
+  moduleName?: string;
+  defaultArgs?: DefaultArguments
+}): TemplateOnlyComponent;
+```
+
+This API will allow us to incorporate future changes and additions to the
+`templateOnly` definition without needing to worry about the order becoming
+unintuitive or difficult to learn. If we wish to save bytes here, the object
+can be compiled down to a smaller representation in the future.
+
+#### Extending Glimmer components
+
+Glimmer components, as noted above, do not have a way to define default
+arguments today, so we propose adding an temporary API that makes them capable
+of doing so.
+
+```ts
+function setDefaultArgs(
+  definition: object,
+  defaultArgs: Record<string, DefaultArgumentValue>
+): void;
+```
+
+Since Glimmer components can only receive named arguments, the function will
+only require that users associate an object containing named argument defaults.
+
+This function will be importable from `@glimmer/component`. It will work _only_
+with Glimmer components. For custom managers provided by addons, they
+should define their own APIs for associating default args. For Classic
+components, users can provide default arguments on the class definition as they
+could historically.
+
+The `defaultArgs` object associated with a component class will be inherited
+by all subclasses of that component, similar to how manager and template
+inheritance works today.
+
+## How we teach this
+
+Managers are generally only described in the API documentation, and the new
+additions to existing components are meant to be low-level compile targets. As
+such, the only additions to documentation will be API docs.
+
+### API Docs for component managers
+
+The following API docs for the `validateArgs` hook should be integrated into the
+API docs for component managers.
+
+#### `getDefaultArgs`
+
+The `getDefaultArgs` hook is an optional hook on component managers. The hook
+receives the component definition as its first argument, and it can return an
+object containing default argument values for the component.
+
+```js
+class MyManager {
+  getDefaultArgs(definition) {
+    return {
+      named: {
+        a: 1,
+        b: () => [],
+      },
+      positional: ['foo']
```

Comment:
Do Glimmer components and their component managers support positional params?

Reply:
Glimmer components do not support positional parameters, which is why the proposed `setDefaultArgs` API for Glimmer components only accepts named arg defaults. In general though, the VM does not really distinguish between components with and without positional parameters. You can invoke a Glimmer component with positional parameters, like so:

```hbs
{{my-glimmer-component @positional1}}
```

And the VM will pass that positional parameter to the component - there is just no way for the component to use that positional argument or refer to it in any way, since the underlying manager does not expose it. That's why in general, manager APIs receive both named and positional args.